common/src/IO/SprParser.cpp

```diff
@@ -115,29 +115,121 @@ static Assets::Orientation parseSpriteOrientationType(Reader& reader) {
   return static_cast<Assets::Orientation>(type);
 }
 
+/**
+ * Specifies the render mode for a Goldsource sprite.
+ * Affects the palette data.
+ */
+enum class RenderMode : int32_t
+{
+  /** No alpha channel, just plain RGB */
+  Normal = 0,
+  /** Normal but also R+G+B/3 is the alpha channel */
+  Additive = 1,
+  /** Index 255 is the colour, R+G+B/3 is the alpha channel */
+  IndexAlpha = 2,
+  /** Standard Quake behaviour, Palette index 255 = transparent */
+  AlphaTest = 3
+};
+
+static RenderMode parseSpriteRenderMode(Reader& reader) {
+  const auto mode = reader.readInt<int32_t>();
+  if (mode < 0 || mode > 3) {
+    throw AssetException{"Unknown SPR render mode: " + std::to_string(mode)};
+  }
+
+  return static_cast<RenderMode>(mode);
+}
+
+static std::vector<unsigned char> processGoldsourcePalette(
+  const RenderMode mode, const std::vector<unsigned char>& data) {
+  // Convert the data into a Goldsource palette
+  auto processed = std::vector<unsigned char>();
+  processed.reserve(1024);
+
+  for (size_t i = 0; i < 256; ++i) {
+    const auto r = data[3 * i + 0];
+    const auto g = data[3 * i + 1];
+    const auto b = data[3 * i + 2];
+
+    // add the RGB channels - for IndexAlpha, the RGB is always index 255
+    if (mode == RenderMode::IndexAlpha) {
+      processed.push_back(data[0xFF * 3 + 0]);
+      processed.push_back(data[0xFF * 3 + 1]);
+      processed.push_back(data[0xFF * 3 + 2]);
+    } else {
+      processed.push_back(r);
+      processed.push_back(g);
+      processed.push_back(b);
+    }
+
+    // Add the alpha channel
+    switch (mode) {
+      case RenderMode::Normal:
+        processed.push_back(0xFF);
+        break;
+      case RenderMode::Additive:
+      case RenderMode::IndexAlpha: {
+        const auto average = round(static_cast<float>(r + g + b) / 3.0f);
+        processed.push_back(static_cast<unsigned char>(average));
+        break;
+      }
+      case RenderMode::AlphaTest:
+        processed.push_back(static_cast<unsigned char>(i == 0xFF ? 0 : 0xFF));
+        break;
+    }
+  }
+
+  return processed;
+}
+
 std::unique_ptr<Assets::EntityModel> SprParser::doInitializeModel(Logger& /* logger */) {
   // see https://www.gamers.org/dEngine/quake/spec/quake-spec34/qkspec_6.htm#CSPRF
 
+  // Half-Life sprites (SPR version 2) are the same as Quake sprites, except
+  // there is an additional integer in the header (render mode), and the palette
+  // data is embedded after the header instead of using the external palette file.
+
   auto reader = m_reader;
 
   const auto ident = reader.readString(4);
   if (ident != "IDSP") {
     throw AssetException{"Unknown SPR ident: " + ident};
   }
 
+  // Version 1: Quake SPR format
+  // Version 2: Half-Life SPR format
   const auto version = reader.readInt<int32_t>();
-  if (version != 1) {
+  if (version != 1 && version != 2) {
     throw AssetException{"Unknown SPR version: " + std::to_string(version)};
   }
 
+  Assets::Palette palette = m_palette;
```

Comment:
I find this control flow too confusing. Can you maybe extract a function `getPalette` that returns the palette and that encapsulates the decision whether to use the supplied or the embedded palette?

Reply:
Would this be sufficient? It avoids needing to pass `m_palette` & `version` through to the `readEmbeddedPalette` method.

```cpp
  Assets::Palette palette = m_palette;
  if (version == 2) {
    palette = readEmbeddedPalette(reader, renderMode);
  }
```

Reply:
One last thing: These helper functions are usually called `parse...`, is there a reason you called this `read...`?

Reply:
No particular reason aside from my own perception of parsing usually involving text of some sort. I've renamed it.

Reply:
Yeah, agree the naming isn't super accurate, but I think it's important to be consistent.